File: core/Common/Utils/AesHelper.cs
// Copyright (C) 2026 Tingyang Zhang
//
// This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License along with this program. If not, see https://www.gnu.org/licenses/.

﻿using System;
using System.IO;
using System.Security.Cryptography;

namespace Nproj.StillHereApp.Common.Utils
{
    /// <summary>
    /// AES加密解密帮助类
    /// </summary>
    public class AesHelper
    {
        /// <summary>
        /// 默认密钥
        /// </summary>
        private static string _key = ConfigHelper.Get(ConfigHelper.AesKey);

        /// <summary>
        /// 默认IV
        /// </summary>
        private static byte[] _iv = new byte[16]
        {
            1, 2, 3, 4, 5, 6, 7, 8,
            9, 10, 11, 12, 13, 14, 15, 16
        };

        /// <summary>  
        /// AES加密  
        /// </summary>  
        /// <param name="str">需要加密字符串</param>  
        /// <returns>加密后字符串</returns>  
        public static string Encrypt(string str)
        {
            return Encrypt(str, _key);
        }

        /// <summary>  
        /// AES解密  
        /// </summary>  
        /// <param name="str">需要解密字符串</param>  
        /// <returns>解密后字符串</returns>  
        public static string Decrypt(string str)
        {
            return Decrypt(str, _key);
        }

        /// <summary>
        /// Hex字符串转字节数组
        /// </summary>
        /// <param name="hex">Hex字符串</param>
        /// <returns></returns>
        private static byte[] HexStringToBytes(string hex)
        {
            if (string.IsNullOrWhiteSpace(hex))
            {
                return Array.Empty<byte>();
            }

            // 去掉空格
            hex = hex.Replace(" ", "");

            if (hex.Length % 2 != 0)
            {
                throw new ArgumentException("Hex string must have even length.");
            }

            var bytes = new byte[hex.Length / 2];
            for (var i = 0; i < hex.Length; i += 2)
            {
                // 每两个字符转换为一个字节
                var byteValue = hex.Substring(i, 2);
                bytes[i / 2] = Convert.ToByte(byteValue, 16);
            }

            return bytes;
        }

        /// <summary>
        /// AES加密
        /// </summary>
        /// <param name="str">需要加密的字符串</param>
        /// <param name="key">32位密钥</param>
        /// <returns>加密后的字符串</returns>
        private static string Encrypt(string str, string key)
        {
            byte[] encrypted;
            var keyArray = HexStringToBytes(key);
            using (var aes = Aes.Create())
            {
                aes.Key = keyArray;
                aes.IV = _iv;
                using var encryptor = aes.CreateEncryptor(aes.Key, aes.IV);
                using (var msEncrypt = new MemoryStream())
                {
                    using (var csEncrypt = new CryptoStream(msEncrypt, encryptor, CryptoStreamMode.Write))
                    {
                        using (var swEncrypt = new StreamWriter(csEncrypt))
                        {
                            swEncrypt.Write(str);
                        }

                        encrypted = msEncrypt.ToArray();
                    }
                }
            }

            return Convert.ToBase64String(encrypted);
        }

        /// <summary>
        /// AES解密
        /// </summary>
        /// <param name="str">需要解密的字符串</param>
        /// <param name="key">32位密钥</param>
        /// <returns>解密后的字符串</returns>
        private static string Decrypt(string str, string key)
        {
            var keyArray = HexStringToBytes(key);
            var strArray = Convert.FromBase64String(str);

            string res = null;
            using var aes = Aes.Create();
            aes.Key = keyArray;
            aes.IV = _iv;

            var decrypt = aes.CreateDecryptor(aes.Key, aes.IV);
            using var msDecrypt = new MemoryStream(strArray);
            using var csDecrypt = new CryptoStream(msDecrypt, decrypt, CryptoStreamMode.Read);
            using var srDecrypt = new StreamReader(csDecrypt);
            res = srDecrypt.ReadToEnd();

            return res;
        }

        /// <summary>
        /// TryDecrypt
        /// </summary>
        /// <param name="str"></param>
        /// <returns></returns>
        public static string TryDecrypt(string str)
        {
            try
            {
                return Decrypt(str);
            }
            catch (Exception e)
            {
                LogHelper.Error($"AesHelper TryDecrypt Error, Msg={e.Message},Str={str}");
                return "";
            }
        }

        /// <summary>
        /// DecryptSignKey
        /// </summary>
        /// <param name="ciphertext"></param>
        /// <param name="userAgent"></param>
        /// <returns></returns>
        public static string DecryptSignKey(string ciphertext, string userAgent)
        {
            if (string.IsNullOrEmpty(ciphertext))
            {
                return "";
            }

            var result = TryDecrypt(ciphertext);

            if (string.IsNullOrEmpty(result))
            {
                return "";
            }

            var arr = result.Split(':');

            if (arr.Length != 2)
            {
                return "";
            }

            var currentTimestamp = SystemHelper.ConvertDateTimeLong(DateTime.Now);
            var timestampConvertResult = long.TryParse(arr[0], out var requestTimestamp);
            var timestampDiff = Math.Abs(currentTimestamp - requestTimestamp);
            var flag = timestampDiff > 120 * 60 * 1000; //一个动态key最长120分钟有效期
            if (flag || !timestampConvertResult)
            {
                return "";
            }

            //0 time
            //1 rand
            return SystemHelper.Md5(userAgent + arr[0] + arr[1]);
        }

        /// <summary>
        /// GenerateSignKey
        /// </summary>
        /// <param name="ciphertext"></param>
        /// <param name="userAgent"></param>
        /// <param name="timestamp"></param>
        /// <param name="random"></param>
        /// <returns></returns>
        public static string GenerateSignKey(out string ciphertext, string userAgent, string timestamp, string random)
        {
            ciphertext = Encrypt(timestamp + ":" + random);
            return SystemHelper.Md5(userAgent + timestamp + random);
        }

        /// <summary>
        /// CheckSign
        /// </summary>
        /// <param name="sign"></param>
        /// <param name="key"></param>
        /// <param name="timestamp"></param>
        /// <param name="random"></param>
        /// <returns></returns>
        public static bool CheckSign(string sign, string key, string timestamp, string random)
        {
            return sign.ToLower() == SystemHelper.Md5(key + "---" + timestamp + "---" + random).ToLower();
        }

        /// <summary>
        /// DecryptUserAgent
        /// </summary>
        /// <param name="userAgent"></param>
        /// <returns></returns>
        public static string DecryptUserAgent(string userAgent)
        {
            userAgent = SystemHelper.UrlDecode(userAgent);
            var str = TryDecrypt(userAgent);
            //将str按':'分割，取第一个部分作为结果返回
            var arr = str.Split(':');
            if (arr.Length > 0)
            {
                return arr[0];
            }

            return "";
        }
    }
}

File: android/app/src/main/java/net/nproj/stillhere/utils/AesHelper.java
/*
 * Copyright (C) 2026 Tingyang Zhang
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with this program. If not, see https://www.gnu.org/licenses/.
 */

package net.nproj.stillhere.utils;

import android.util.Base64;

import java.nio.charset.StandardCharsets;

import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

/**
 * AES 加密解密工具类
 */
public class AesHelper {

    /**
     * AES 密钥（16、24、32 字节分别对应 AES-128、AES-192、AES-256）
     */
    private static final String KEY = "4A9C8B2D5E1F0A3B6C9D8E7F1A2B3C4D5E6F7A8B9C0D1E2F"; //AES-192

    /**
     * 初始化向量（16 字节）
     */
    private static final byte[] IV = new byte[]{
            1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16
    };

    /**
     * 加密转换模式
     */
    private static final String TRANSFORMATION = "AES/CBC/PKCS5Padding";

    /**
     * 加密算法
     */
    private static final String ALGORITHM = "AES";

    /**
     * 将十六进制字符串转换为字节数组
     *
     * @param hex 十六进制字符串
     * @return 字节数组
     */
    private static byte[] hexStringToBytes(String hex) {
        if (hex == null || hex.isEmpty()) {
            return new byte[0];
        }

        // 去掉可能的空格
        hex = hex.replaceAll("\\s", "");

        int len = hex.length();
        if (len % 2 != 0) {
            throw new IllegalArgumentException("Invalid hex string, length must be even");
        }

        byte[] result = new byte[len / 2];

        for (int i = 0; i < len; i += 2) {
            // 每两个字符表示一个字节
            String byteStr = hex.substring(i, i + 2);
            result[i / 2] = (byte) Integer.parseInt(byteStr, 16);
        }

        return result;
    }

    /**
     * AES 加密
     *
     * @param plainText 原始明文
     * @return Base64 编码的密文
     */
    public static String encrypt(String plainText) {
        try {
            Cipher cipher = Cipher.getInstance(TRANSFORMATION);
            SecretKeySpec keySpec = new SecretKeySpec(hexStringToBytes(KEY), ALGORITHM);
            IvParameterSpec ivSpec = new IvParameterSpec(IV);
            cipher.init(Cipher.ENCRYPT_MODE, keySpec, ivSpec);

            byte[] encrypted = cipher.doFinal(plainText.getBytes(StandardCharsets.UTF_8));
            return Base64.encodeToString(encrypted, Base64.NO_WRAP);
        } catch (Exception e) {
            //throw new RuntimeException("AES 加密失败", e);
            LogHelper.Error("AES 加密失败: " + e.getMessage());
            return "";
        }
    }


    /**
     * AES 解密
     *
     * @param cipherText Base64 编码的密文
     * @return 解密后的明文
     */
    public static String decrypt(String cipherText) {
        try {
            Cipher cipher = Cipher.getInstance(TRANSFORMATION);
            SecretKeySpec keySpec = new SecretKeySpec(hexStringToBytes(KEY), ALGORITHM);
            IvParameterSpec ivSpec = new IvParameterSpec(IV);
            cipher.init(Cipher.DECRYPT_MODE, keySpec, ivSpec);

            //byte[] decoded = Base64.getDecoder().decode(cipherText);
            byte[] decoded = Base64.decode(cipherText, Base64.NO_WRAP);
            byte[] decrypted = cipher.doFinal(decoded);
            return new String(decrypted, StandardCharsets.UTF_8);
        } catch (Exception e) {
            //throw new RuntimeException("AES 解密失败", e);
            return "";
        }
    }

    /**
     * 获取一个加了密的Android ID 这个东西只能用一次
     *
     * @return 加密后的Android ID
     */
    public static String GetAndroidId() {
        var str = GlobalConstHelper.ANDROID_ID + ":" + Helper.getGuid() + ":" + Helper.getTimestampMs();
        return AesHelper.encrypt(str);
    }
}
